File: 39주차/실패율.py
def solution(n,stages):
    answer = [[]]

    arr = [0]*(n+2)
    passUser = len(stages)# 지금 스테이지까지 도달한 사용자 수(도전자 포함.)
    for v in stages:
        arr[v] += 1
    for i in range(1,n+1):
        if passUser > 0:
            rate = arr[i]/passUser # 0이 분모가 될 수 있음에 유의
        else:
            rate=0
        answer.append([i,rate])
        passUser -= arr[i]
    answer = answer[1:]
    answer.sort(key=lambda x:(-x[1],x[0]))


    return [v[0] for v in answer]
# 아직 아무도 도달하지 않은 스테이지라면 실패율은 0.
# 프로그래머스 실패율.

File: 51주차/실패율.js
function solution(N, stages) {
  var answer = []; // [실패율, 스테이지 번호]형식.

  let total = [0];
  let sum = 0;
  let arr = Array(N + 2).fill(0);
  // 각 스테이지 별로 몇개 있는지 체크.
  stages.forEach((v) => (arr[v] += 1));

  // 총 합을 더해서 sum에 push
  for (let i = 1; i < arr.length; i++) {
    sum += arr[i];
    total.push(sum);
  }
  // total[end] - total[i-1] => 현재 스테이지에 도달한 플레이이어 수
  // arr[i] => 현재 스테이지에 도달했지만, 클리어못한 수.
  const end = total[total.length - 1];
  for (let i = 1; i < total.length - 1; i++) {
    answer.push([arr[i] / (end - total[i - 1]), i]);
  }

  answer.sort((a, b) => {
    if (a[0] !== b[0]) return b[0] - a[0];
    else a[1] - b[1];
  });
  answer = answer.map((v) => v[1]);
  return answer;
}
// 프로그래머스.
